crates/prover-tracer/src/lib.rs

```diff
@@ -0,0 +1,283 @@
+                        .unwrap_or(span_limits_default.max_attributes_per_link),
+                    _ => span_limits_default.max_attributes_per_link,
+                })
+                .with_resource(Resource::builder().with_attributes(resources).build())
+                .with_span_processor(
+                    BatchSpanProcessor::builder(otlp_exporter)
+                        .with_batch_config(batch_processor_config.build())
+                        .build(),
+                )
+                .build();
+
+            let tracer = trace_provider.tracer("agglayer-otlp");
+
+            let _ = global::set_tracer_provider(trace_provider);
+
+            layers.push(
+                tracing_opentelemetry::layer()
+                    .with_tracer(tracer)
+                    .with_filter(
+                        EnvFilter::try_from_default_env().unwrap_or_else(|_| config.level.into()),
+                    )
+                    .boxed(),
+            );
+
+            global::set_text_map_propagator(TraceContextPropagator::new());
+        } else {
+            anyhow::bail!("Otlp tracing requires both otlp agent url and otlp service provided");
+        }
+    }
+
+    layers.push(match config.format {
+        TracingFormat::Pretty => tracing_subscriber::fmt::layer()
+            .pretty()
+            .with_writer(writer.as_make_writer())
+            .with_filter(EnvFilter::try_from_default_env().unwrap_or_else(|_| config.level.into()))
+            .boxed(),
+
+        TracingFormat::Json => tracing_subscriber::fmt::layer()
+            .json()
+            .with_writer(writer.as_make_writer())
+            .with_filter(EnvFilter::try_from_default_env().unwrap_or_else(|_| config.level.into()))
+            .boxed(),
+    });
+
+    // We are using try_init because integration test may try
+    // to initialize this multiple times.
+    _ = tracing_subscriber::Registry::default()
+        .with(layers)
+        .try_init();
+
+    tracing::info!("Tracing initialized with config: {config:?}");
+
+    Ok(())
+}
+
+fn build_resources(otlp_service_name: &str, version: &str) -> Vec<KeyValue> {
+    let mut resources = Vec::new();
+
+    resources.push(KeyValue::new("service.name", otlp_service_name.to_string()));
+    resources.push(KeyValue::new("service.version", version.to_string()));
+
+    let custom_resources: Vec<_> = std::env::var("AGGLAYER_OTLP_TAGS")
+        .unwrap_or_default()
+        .split(',')
+        // NOTE: limit to 10 tags to avoid exploit
+        .take(10)
+        .filter_map(|tag_raw| {
+            let mut v = tag_raw.splitn(2, '=');
+            match (v.next(), v.next()) {
+                (Some(key), Some(value)) if !key.trim().is_empty() && !value.trim().is_empty() => {
+                    Some(KeyValue::new(
+                        key.trim().to_string(),
+                        value.trim().to_string(),
+                    ))
+                }
+                _ => None,
+            }
```

Comment:
I think you could use `split_once` here:
```suggestion
            let (key, val) = tag_raw.split_once(2, '=')?;
            let key = key.trim();
            let val = val.trim();
            if key.is_empty() || val.is_empty() {
                return None;
            }
            Some((key.to_string(), val.to_string()))
```